source/http2_stream_manager.c

```diff
@@ -0,0 +1,987 @@
+                    chosen_connection->max_concurrent_streams);
+            }
+        }
+    }
+    AWS_ASSERT(errored == 0 && "random access set went wrong");
+}
+
+/* NOTE: never invoke with lock held */
+static void s_finish_pending_acquisitions_list_helper(
+    struct aws_http2_stream_manager *stream_manager,
+    struct aws_linked_list *pending_acquisitions,
+    int error_code) {
+    while (!aws_linked_list_empty(pending_acquisitions)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(pending_acquisitions);
+        struct aws_h2_sm_pending_stream_acquisition *pending_acquisition =
+            AWS_CONTAINER_OF(node, struct aws_h2_sm_pending_stream_acquisition, node);
+        /* Make sure no connection assigned. */
+        AWS_ASSERT(pending_acquisition->sm_connection == NULL);
+        if (pending_acquisition->callback) {
+            pending_acquisition->callback(NULL, error_code, pending_acquisition->user_data);
+        }
+        STREAM_MANAGER_LOGF(
+            DEBUG,
+            stream_manager,
+            "acquisition:%p failed with error: %d(%s)",
+            (void *)pending_acquisition,
+            error_code,
+            aws_error_str(error_code));
+        s_pending_stream_acquisition_destroy(pending_acquisition);
+    }
+}
+
+/* This is scheduled to run on a separate event loop to finish pending acquisition asynchronously */
+static void s_finish_pending_acquisitions_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)status;
+    struct aws_http2_stream_manager *stream_manager = arg;
+    STREAM_MANAGER_LOG(TRACE, stream_manager, "Stream Manager final task runs");
+    struct aws_http2_stream_management_transaction work;
+    struct aws_linked_list pending_acquisitions;
+    aws_linked_list_init(&pending_acquisitions);
+    s_aws_stream_management_transaction_init(&work, stream_manager);
+    { /* BEGIN CRITICAL SECTION */
+        s_lock_synced_data(stream_manager);
+        AWS_ASSERT(stream_manager->synced_data.state == AWS_H2SMST_DESTROYING);
+        /* swap list to avoid callback with lock held. */
+        aws_linked_list_swap_contents(&pending_acquisitions, &stream_manager->synced_data.pending_acquisitions);
+        /* After the callbacks invoked, now we can update the count */
+        s_sm_count_decrease_synced(
+            stream_manager, AWS_SMCT_PENDING_ACQUISITION, stream_manager->synced_data.pending_acquisition_count);
+        s_aws_http2_stream_manager_build_transaction_synced(&work);
+        s_unlock_synced_data(stream_manager);
+    } /* END CRITICAL SECTION */
+    s_finish_pending_acquisitions_list_helper(
+        stream_manager, &pending_acquisitions, AWS_ERROR_HTTP_STREAM_MANAGER_SHUTTING_DOWN);
+    aws_mem_release(stream_manager->allocator, task);
+    s_aws_http2_stream_manager_execute_transaction(&work);
+}
+
+static void s_check_new_connections_needed_synced(struct aws_http2_stream_management_transaction *work) {
```

Comment:
comment  like:
/* helper function for building the transaction
sets how many new connections we should request 
*/

and maybe comment on other helper functions like it - and/or common naming pattern